README.md
HTTP Strict Transport Security middleware
========================================
[![Build Status](https://travis-ci.org/helmetjs/hsts.svg?branch=master)](https://travis-ci.org/helmetjs/hsts)

This middleware adds the `Strict-Transport-Security` header to the response. This tells browsers, "hey, only use HTTPS for the next period of time". ([See the spec](http://tools.ietf.org/html/rfc6797) for more.) Note that the header won't tell users on HTTP to *switch* to HTTPS, it will just tell HTTPS users to stick around. You can enforce HTTPS with the [express-enforces-ssl](https://github.com/aredo/express-enforces-ssl) module.

This will set the Strict Transport Security header, telling browsers to visit by HTTPS for the next 180 days:

```javascript
const hsts = require('hsts')

app.use(hsts({
  maxAge: 15552000  // 180 days in seconds
}))
// Strict-Transport-Security: max-age: 15552000; includeSubDomains
```

Note that the max age must be in seconds.

The `includeSubDomains` directive is present by default. If this header is set on *example.com*, supported browsers will also use HTTPS on *my-subdomain.example.com*. You can disable this:

```javascript
app.use(hsts({
  maxAge: 15552000,
  includeSubDomains: false
}))
```

Some browsers let you submit your site's HSTS to be baked into the browser. You can add `preload` to the header with the following code. You can check your eligibility and submit your site at [hstspreload.org](https://hstspreload.org/).

```javascript
app.use(hsts({
  maxAge: 31536000,        // Must be at least 1 year to be approved
  includeSubDomains: true, // Must be enabled to be approved
  preload: true
}))
```

This header will always be set because [the header is ignored in insecure HTTP](https://tools.ietf.org/html/rfc6797#section-8.1). You may wish to set it conditionally:

```javascript
const hstsMiddleware = hsts({
  maxAge: 1234000
})

app.use((req, res, next) => {
  if (req.secure) {
    hstsMiddleware(req, res, next)
  } else {
    next()
  }
})
```

This header is [somewhat well-supported by browsers](https://caniuse.com/#feat=stricttransportsecurity).


CHANGELOG.md
# Changelog

## Unreleased
### Added
* TypeScript type definitions. See [#25](https://github.com/helmetjs/hsts/pull/25)
* `maxAge` now supports `BigInt`s

### Changed
* Dropped support for Node <10
* `maxAge` must now be a number if supplied; `undefined` and `null` are no longer allowed
* Removed deprecated `setIf` option. See [#22](https://github.com/helmetjs/hsts/issues/22)
* Removed deprecated `includeSubdomains` option; use `includeSubDomains` instead. See [#21](https://github.com/helmetjs/hsts/issues/21)

## 2.2.0 - 2019-03-10
### Added
* Created a changelog

### Changed
* Mark the module as Node 4+ in the `engines` field of `package.json`
* Add a `homepage` in `package.json`
* Add an email to `package.json`'s `bugs` field
* Updated documentation
* Updated Adam Baldwin's contact info. See [helmetjs/helmet#189](https://github.com/helmetjs/helmet/issues/189)

### Deprecated
* The `setIf` option has been deprecated and will be removed in `hsts@3`. Refer to the documentation to see how to do without it. See [#22](https://github.com/helmetjs/hsts/issues/22) for more
* The `includeSubdomains` option (with a lowercase `d`) has been deprecated and will be removed in `hsts@3`. Use the uppercase-D `includeSubDomains` option instead. See [#21](https://github.com/helmetjs/hsts/issues/21) for more

Changes in versions 2.1.0 and below can be found in [Helmet's changelog](https://github.com/helmetjs/helmet/blob/master/CHANGELOG.md).


test/index.js
var hsts = require('..')

var assert = require('assert')
var connect = require('connect')
var request = require('supertest')

describe('hsts', function () {
  function app (options) {
    var result = connect()
    result.use(hsts(options))
    result.use(function (req, res) {
      res.end('Hello world!')
    })
    return result
  }

  it('throws an error with invalid parameters', function () {
    assert.throws(hsts.bind(this, { maxAge: -123 }))
    assert.throws(hsts.bind(this, { maxAge: '123' }))
    assert.throws(hsts.bind(this, { maxAge: true }))
    assert.throws(hsts.bind(this, { maxAge: false }))
    assert.throws(hsts.bind(this, { maxAge: {} }))
    assert.throws(hsts.bind(this, { maxAge: [] }))

    assert.throws(hsts.bind(this, { maxAge: 1234 }, 'extra argument'))

    assert.throws(hsts.bind(this, { maxage: false }))
    assert.throws(hsts.bind(this, { maxage: 1234 }))
  })

  it('by default, sets max-age to 180 days and adds "includeSubDomains"', function () {
    assert.strictEqual(15552000, 180 * 24 * 60 * 60)

    return request(app())
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=15552000; includeSubDomains')
  })

  it('can set max-age to a positive integer', function () {
    return request(app({
      maxAge: 1234
    }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=1234; includeSubDomains')
  })

  it('rounds the max-age', function () {
    return request(app({
      maxAge: 1234.56
    }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=1235; includeSubDomains')
  })

  it('can set max-age to -0', function () {
    return request(app({
      maxAge: -0
    }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=0; includeSubDomains')
  })

  it('can set max-age to 0', function () {
    return request(app({
      maxAge: 0
    }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=0; includeSubDomains')
  })

  it('can disable subdomains with the includeSubDomains option', function () {
    return request(app({
      includeSubDomains: false
    }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=15552000')
  })

  it('can enable preloading', function () {
    return request(app({
      preload: true
    }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=15552000; includeSubDomains; preload')
  })

  it('does nothing with the `force` option; allowed for backwards compatibility', function () {
    // We should remove this test in hsts@3.
    return request(app({ force: true }))
      .get('/')
      .expect(200)
      .expect('Strict-Transport-Security', 'max-age=15552000; includeSubDomains')
  })

  it('names its function and middleware', function () {
    assert.strictEqual(hsts.name, 'hsts')
    assert.strictEqual(hsts().name, 'hsts')
  })
})


test/index.test.ts
import connect from 'connect';
import request from 'supertest';
import { IncomingMessage, ServerResponse } from 'http';

import hsts = require('..');

describe('hsts', () => {
  function app(middleware: ReturnType<typeof hsts>) {
    const result = connect();
    result.use(middleware);
    result.use((_req: IncomingMessage, res: ServerResponse) => {
      res.end('Hello world!');
    });
    return result;
  }

  it('throws an error with invalid parameters', () => {
    /* eslint-disable @typescript-eslint/no-explicit-any */
    expect(() => hsts({ maxAge: -123 })).toThrow();
    expect(() => hsts({ maxAge: BigInt(-123) })).toThrow();
    expect(() => hsts({ maxAge: -0.1 })).toThrow();
    expect(() => hsts({ maxAge: Infinity })).toThrow();
    expect(() => hsts({ maxAge: -Infinity })).toThrow();
    expect(() => hsts({ maxAge: NaN })).toThrow();

    expect(() => hsts({ maxAge: '123' } as any)).toThrow();
    expect(() => hsts({ maxAge: true } as any)).toThrow();
    expect(() => hsts({ maxAge: false } as any)).toThrow();
    expect(() => hsts({ maxAge: {} } as any)).toThrow();
    expect(() => hsts({ maxAge: [] } as any)).toThrow();
    expect(() => hsts({ maxAge: null } as any)).toThrow();
    expect(() => hsts({ maxAge: undefined } as any)).toThrow();

    expect(() => hsts({ maxage: false } as any)).toThrow();
    expect(() => hsts({ maxage: 1234 } as any)).toThrow();

    expect(() => hsts({ includeSubdomains: false } as any)).toThrow();
    expect(() => hsts({ includeSubdomains: true } as any)).toThrow();
    /* eslint-enable @typescript-eslint/no-explicit-any */
  });

  it('by default, sets max-age to 180 days and adds "includeSubDomains"', async () => {
    expect(15552000).toStrictEqual(180 * 24 * 60 * 60);

    await request(app(hsts()))
      .get('/')
      .expect(
        'Strict-Transport-Security',
        'max-age=15552000; includeSubDomains'
      );
  });

  it('can set max-age to a positive integer', async () => {
    await request(app(hsts({
      maxAge: 1234,
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=1234; includeSubDomains');
  });

  it('rounds the max-age', async () => {
    await request(app(hsts({
      maxAge: 1234.56,
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=1235; includeSubDomains');

    await request(app(hsts({
      maxAge: 1234.49,
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=1234; includeSubDomains');
  });

  it('can set max-age to -0', async () => {
    await request(app(hsts({
      maxAge: -0,
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=0; includeSubDomains');
  });

  it('can set max-age to 0', async () => {
    await request(app(hsts({
      maxAge: 0,
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=0; includeSubDomains');
  });

  it('can set max-age to a bigint', async () => {
    await request(app(hsts({
      maxAge: BigInt(1234),
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=1234; includeSubDomains');
  });

  it('can disable subdomains with the includeSubDomains option', async () => {
    await request(app(hsts({
      includeSubDomains: false,
    })))
      .get('/')
      .expect('Strict-Transport-Security', 'max-age=15552000');
  });

  it('can enable preloading', async () => {
    await request(app(hsts({
      preload: true,
    })))
      .get('/')
      .expect(
        'Strict-Transport-Security',
        'max-age=15552000; includeSubDomains; preload'
      );
  });

  it('names its function and middleware', () => {
    expect(hsts.name).toStrictEqual('hsts');
    expect(hsts().name).toStrictEqual('hsts');
  });
});


index.js
var DEFAULT_MAX_AGE = 180 * 24 * 60 * 60

module.exports = function hsts (options) {
  options = options || {}

  var maxAge = options.maxAge != null ? options.maxAge : DEFAULT_MAX_AGE
  var includeSubDomains = options.includeSubDomains !== false

  if (options.hasOwnProperty('maxage')) {
    throw new Error('maxage is not a supported property. Did you mean to pass "maxAge" instead of "maxage"?')
  }
  if (arguments.length > 1) {
    throw new Error('HSTS passed the wrong number of arguments.')
  }
  if (typeof maxAge !== 'number') {
    throw new TypeError('HSTS must be passed a numeric maxAge parameter.')
  }
  if (maxAge < 0) {
    throw new RangeError('HSTS maxAge must be nonnegative.')
  }

  var header = 'max-age=' + Math.round(maxAge)
  if (includeSubDomains) {
    header += '; includeSubDomains'
  }
  if (options.preload) {
    header += '; preload'
  }

  return function hsts (req, res, next) {
    res.setHeader('Strict-Transport-Security', header)
    next()
  }
}


index.ts
import { IncomingMessage, ServerResponse } from 'http';

interface HstsOptions {
  maxAge?: number | bigint;
  includeSubDomains?: boolean;
  preload?: boolean;
}

function getMaxAgeValue(maxAge: unknown): number | bigint {
  if (typeof maxAge === 'number') {
    if (maxAge < 0) {
      throw new RangeError('HSTS maxAge must be nonnegative.');
    } else if (Number.isNaN(maxAge)) {
      throw new Error('HSTS maxAge must not be NaN.');
    } else if (!Number.isFinite(maxAge)) {
      throw new Error('HSTS maxAge must be a finite number.');
    } else {
      return Math.round(maxAge);
    }
  } else if (typeof maxAge === 'bigint') {
    if (maxAge < 0) {
      throw new RangeError('HSTS maxAge must be nonnegative.');
    } else {
      return maxAge;
    }
  } else {
    throw new TypeError('HSTS must be passed a numeric maxAge parameter.');
  }
}

function getHeaderValueFromOptions(options: HstsOptions): string {
  const DEFAULT_MAX_AGE = 180 * 24 * 60 * 60;

  if ('maxage' in options) {
    throw new Error('"maxage" is not a supported property. Did you mean to use "maxAge" instead?');
  }
  if ('includeSubdomains' in options) {
    throw new Error('"includeSubdomains" is not a supported property. Did you mean to use "includeSubDomains" instead?');
  }

  const maxAge = 'maxAge' in options ? options.maxAge : DEFAULT_MAX_AGE;
  let header = `max-age=${getMaxAgeValue(maxAge)}`;
  if (options.includeSubDomains !== false) {
    header += '; includeSubDomains';
  }
  if (options.preload) {
    header += '; preload';
  }

  return header;
}

export = function hsts(options: HstsOptions = {}) {
  const headerValue = getHeaderValueFromOptions(options);

  return function hsts(
    _req: IncomingMessage,
    res: ServerResponse,
    next: () => void
  ) {
    res.setHeader('Strict-Transport-Security', headerValue);
    next();
  };
};
